ui/apps/platform/src/Containers/Vulnerabilities/hooks/useInvalidateVulnerabilityQueries.ts

```diff
@@ -0,0 +1,74 @@
+import { useApolloClient } from '@apollo/client';
+import { DocumentNode, FieldNode, OperationDefinitionNode } from 'graphql';
+import uniq from 'lodash/uniq';
+
+import { imageListQuery } from '../WorkloadCves/Tables/ImagesTable';
+import { cveListQuery } from '../WorkloadCves/Tables/CVEsTable';
+import { deploymentListQuery } from '../WorkloadCves/Tables/DeploymentsTable';
+import { deploymentVulnerabilitiesQuery } from '../WorkloadCves/Deployment/DeploymentPageVulnerabilities';
+import { imageVulnerabilitiesQuery } from '../WorkloadCves/Image/ImagePageVulnerabilities';
+import {
+    imageCveAffectedDeploymentsQuery,
+    imageCveAffectedImagesQuery,
+} from '../WorkloadCves/ImageCve/ImageCvePage';
+
+/**
+ * Extracts the top-level field names from a GraphQL document node. This is used to determine
+ * which query fields to invalidate.
+ *
+ * @param documentNode A GraphQL document node
+ * @returns An array of top-level field names
+ */
+function extractTopLevelFieldNames(documentNode: DocumentNode): string[] {
+    return documentNode.definitions
+        .filter((def): def is OperationDefinitionNode => def.kind === 'OperationDefinition')
+        .filter((def) => def.operation === 'query')
+        .flatMap((def) =>
+            def.selectionSet.selections
+                .filter((sel): sel is FieldNode => sel.kind === 'Field')
+                .map((sel) => sel.name.value)
+        );
+}
+
+const queryFieldNames = uniq(
+    [
+        // Queries for the tables on the Workload CVE list page
+        imageListQuery,
+        cveListQuery,
+        deploymentListQuery,
+        // Queries for the Workload CVE detail page
+        imageCveAffectedDeploymentsQuery,
+        imageCveAffectedImagesQuery,
+        // Query for the Workload CVE image detail page
+        imageVulnerabilitiesQuery,
+        // Query for the Workload CVE deployment detail page
+        deploymentVulnerabilitiesQuery,
+    ].flatMap(extractTopLevelFieldNames)
+);
```

Comment:
This entire section can be replaced with
```suggestion
const queryFieldNames = ['image', 'images', 'imageCVE', 'imageCVEs', 'deployment', 'deployments'];
```
and have identical functionality.

The current version explicitly imports the queries used throughout Workload CVEs and extracts the top level fields of each query, which we then evict from the cache. The idea was to tie the dependencies to the queries that are actually in use instead of using "magic strings", but it does add a lot of boilerplate to the code. If we went the "magic strings" route, another bonus is that the code that calls this hook would be simpler since there would no longer be a circular dependency.

Reply:
Good points. My thoughts on each approach:

1. Using the "magic strings"

This simplifies things, but I'm wary of it because we're no longer using the same source of truth. If we update those queries (change the names), then issues will arise that we may not catch. If we don't imagine changing the names, then it may be acceptable to go this route.

2. Using the "extractTopLevelFieldNames" method

I like this approach, and it makes it more reliable that we invalidate the right queries because we're using the same source of truth. I was concerned by the circular dependency though. It defines these queries in the Components. Would it help if we separated the queries into different files?

Reply:
For 1. - I really don't think we will be changing the top level fields any time soon, but if we did, not invalidating those queries after a mutation would be a _very_ subtle bug that would probably wouldn't catch unless we had tests specifically for that.


I'll keep this as the second approach. The only place the circular dependency is a problem is on the image single page, since the action that causes the invalidation happens at the same level that the top level GraphQL query is defined. (Hence the need to pass the invalidation callback down from `ImagePage` instead of using the hook directly due to a possible circular dependency.) 

The way the Workload CVE sections are currently structured is that individual components are declaring data dependencies on a GraphQL fragment, and then the top level component in the route aggregates these fragments into a query to pass down to child components.

The queries themselves are designed to not be reusable. The re-usability comes from the fragment + child-component relationship and the queries are tightly coupled to the place where the data fetching occurs. If we moved the queries into different files, I worry that it might add indirection between pieces that currently have a strong correlation.

Maybe extracting the queries into a custom hook that is colocated with the top level components? This hook would then need to import fragments from whatever child components were used in the parent, which might be a little bit awkward. Although if I was able to get back to https://github.com/stackrox/stackrox/pull/6303 this wouldn't be an issue.